In writeBytes return actual numBytesWritten

hiddenalpha opened the pull request:
This PR originates from https://github.com/java-native/jssc/issues/129 which asks about support to return the actual number of bytes written in place of just a boolean.

Here are some citations from `man 2 write`:

> DESCRIPTION:
> 
> The number of bytes written may be less than count if, for example, there is insufficient space on the underlying physical medium, or the RLIMIT_FSIZE resource limit is encountered (see `man 2 setrlimit`), or the call was interrupted by a signal handler after having written less than count bytes. (See also `man 7 pipe`.)
> 
> RETURN VALUE:
> On success, the number of bytes written is returned.
> 
> Note that a successful write() may transfer fewer than count bytes. Such partial writes can occur for various reasons; for example, because there was insufficient space on the disk device to write all of the requested bytes, or because a blocked write() to a socket, pipe, or similar was interrupted by a signal handler after it had transferred some, but before it had transferred all of the requested bytes. In the event of a partial write, the caller can make another write() call to transfer the remaining bytes. The subsequent call will either transfer further bytes or may result in an error (e.g., if the disk is now full).

Fixes: #129

hiddenalpha committed:
Make writeBytes() return how many bytes actually got written

The core change is this one:

--- src/main/java/jssc/SerialNativeInterface.java
+++ src/main/java/jssc/SerialNativeInterface.java
@@ -278,5 +278,5 @@ public class SerialNativeInterface {
      * @return If the operation is successfully completed, the method returns true, otherwise false
      */
-    public native boolean writeBytes(long handle, byte[] buffer) throws IOException;
+    public native int     writeBytes(long handle, byte[] buffer) throws IOException;

     /**

Essentially changing the return type of `writeBytes()` from `boolean` to
`int`. The reason behind this is that we can return how many bytes
actually got written, in place of just true/false.

Basically, this is a very small change. Nevertheless, there are some
consequences.

This is (kind of) a breaking change for the java API. But still not
really an issue, as it is a very easy migration. About how to migrate,
one can just do the same as done in `SerialPort.writeBytes1()` (notice
the '1' in the name). Just check if returned value matches the length of
the passed-in array. There is also a YAGNI approach (NOT RECOMMENDED!):
One could just use `writeBytes1()` instead. Which provides the old
behavior.

A more severe issue can arise because it potentially also breaks the ABI.
make sure to NOT mix usage of parts of jssc built before or after
this change! in most cases this shouldn't be an issue, as jssc unpacks
the correct version from its JAR anyway. BUT FOR THOSE WHO DID SEPARATE
THE JAVA CODE FROM THE NATIVE CODE AND INSTALL IT SEPARATELY, THEY
POTENTIALLY CAN END UP IN TROUBLE WHEN NOT TAKING CARE HERE.

In addition, this commit also contains some minor cleanups and fixes
some compiler warnings.

Related:
- https://github.com/java-native/jssc/issues/129

(wip @ 17cc247b69cdae978ad7192a21548162fe118f2a)

tresf commented on src/main/cpp/_nix_based/jssc.cpp:
```diff
@@ -62,7 +60,16 @@
 #include "version.h"
 
-//#include <iostream> //-lCstd use for Solaris linker
+#if defined(_MSC_VER) && _MSC_VER < 1800
```
```suggestion
// For snprintf formatting 
#if defined(_MSC_VER) && _MSC_VER < 1800
```

tresf commented on src/main/cpp/windows/jssc.cpp:
```diff
@@ -29,7 +29,16 @@
 #include "version.h"
 
-//#include <iostream>
+#if defined(_MSC_VER) && _MSC_VER < 1800
```
This code is redundant and slightly anti-pattern since it contains conditions which should never exist (e.g. checking for `_MSC_VER < 1800` on Unix).  Would it instead make sense to move this to a common header file?  What's the normal course of action for helpers like this?

hiddenalpha replied:
Thx :) I realized that it was not really needed in "*_nix_based/jssc.cpp*". So I just could remove it there (*duplicate gone*).

> Would it instead make sense to move this to a common header file? What's the normal course of action for helpers like this?

Shared headers for this case I would say. But until now I tried to prevent introducing stuff like that to keep the project simpler overall.

tresf commented on src/main/java/jssc/SerialPort.java:
```diff
@@ -404,8 +406,27 @@ public boolean setDTR(boolean enabled) throws SerialPortException {
+     * @deprecated use {@link #writeBytes(byte[])}.
+     */
+    @Deprecated
+    public boolean writeBytes1(byte[] buffer) throws SerialPortException {
```
My instinct is to remove this (or if we keep it, give it a more self-documenting name).

hiddenalpha replied:
Done :)

I wasn't sure either if I should introduce that method in the first place.

tresf commented on src/main/java/jssc/SerialPort.java:
```diff
@@ -427,7 +448,7 @@ public boolean writeBytes(byte[] buffer) throws SerialPortException {
      */
     public boolean writeByte(byte singleByte) throws SerialPortException {
```
Would it be more consistent to the developers to use `int` everywhere?

hiddenalpha replied:
Pro "change to `int`":
- Makes sense for consistency.
- IMHO API would look more clean.
- I guess a lot of code ignores the return values anyway. So the change won't really hurt.

Pro "keep `boolean`":
- Changing all the signatures will affect a broader surface of jssc's public API. So more consuming code will have to migrate the calls (given they check the return values in the 1st place).
- The answer is not as clear when working with `java.lang.String`s as they may produce more bytes than they contain chars.

I'm not sure which one to choose. Both sound viable. I'll have a look if it's easy to change them aswell :)

**Edit**: Maybe we even need to decide for every function individually? (*eg keep the String/encoding affected ones as bool*?)

tresf commented:
Thanks!  I tried to look into other `write()` functions and had a hard time determining if it was more standard/expected to use `boolean` or `int`.  I understand why a boolean is used for a single byte, but I'm not sure if this is helpful to developers to diverge these two.  I also had some thoughts to share on the precompiler checks for snprintf as it seems like the code is both redundant as well as unused for certain compiler / os combos so may be best moved to an include?  Thoughts welcome.

tresf commented on src/main/cpp/jssc_SerialNativeInterface.h:
```diff
@@ -1,173 +0,0 @@
-/* DO NOT EDIT THIS FILE - it is machine generated */
```
This file is kept for legacy platforms (e.g. ppc, solaris, etc) where leveraging the entire cmake build system is too hard due to outdated dependencies.  Rather than removing it, I would recommend updating it.

tresf commented on src/main/java/jssc/SerialPort.java:
```diff
@@ -435,46 +437,67 @@ public boolean writeByte(byte singleByte) throws SerialPortException {
+     * @return
+     *      Number of bytes written. WARN: Number of BYTES, NOT number of CHARS!
+     *      This may differ from <code>string.length()</code> if passed in string
+     *      contains chars outside the english alphabeth. If you need reliable
```
```suggestion
     *      contains chars outside the english alphabet. If you need reliable
```

tresf commented on src/main/java/jssc/SerialPort.java:
```diff
@@ -435,46 +437,67 @@ public boolean writeByte(byte singleByte) throws SerialPortException {
+     * @return
+     *      Number of bytes written. WARN: Number of BYTES, NOT number of CHARS!
+     *      This may differ from <code>string.length()</code> if passed in string
+     *      contains chars outside the english alphabeth. If you need reliable
```
```suggestion
     *      contains chars outside the english alphabet. If you need reliable
```

hiddenalpha committed:
Apply PR feedback

- Drop legacy function `writeBytes1()`
  https://github.com/java-native/jssc/pull/195#discussion_r2237037444
- Add superfluous comment as requested
  https://github.com/java-native/jssc/pull/195#discussion_r2237009596
- Also change all the other `write..` related calls to `int`
  https://github.com/java-native/jssc/pull/195#discussion_r2237054271

Edit:
- Add back "jssc_SerialNativeInterface.h" again (copy-paste as generated from build)
  https://github.com/java-native/jssc/pull/195#discussion_r2240078225
- Fix typo.
  https://github.com/java-native/jssc/pull/195#discussion_r2240083647
  https://github.com/java-native/jssc/pull/195#discussion_r2240085829

(wip @ dd6dfea3d1e73e1dd0b3585205bb678b9a42286e)

hiddenalpha commented:
Before we merge, I think we should discuss this aspect of the change:

As we change the signature of a native call, one will not be able to use for example a new jar with an old native binary. (*no one should do that anyway*)

So for exmaple if some project decided to install the binary through some other means (*than the auto-unpacking*), like for example compiling it separately and/or installing the binary manually alongside the jar.

This may cause trouble if the project will only update the jar but does not update the binary, it will NOT be compatible. As the new jar will expect numberOfBytesWritten to be returned, but the (old) native binary still will return the boolean.

**Do we need to take care for this case from our side?** Or can we expect projects to properly update both (jar and binary) simultaneously?

If we need to take care of that, I'll need to add additional backward compatibility to the native code.

tresf commented:
> Do we need to take care for this case from our side? Or can we expect projects to properly update both (jar and binary) simultaneously?

We should probably do another minor version bump if we break compat.  I think this is OK.

What I'm still a bit unsure about is the decision to return the written bytes in general.  Since this PR doesn't link a bug report to describe why, it's hard to understand the use-case that this solves.

@hiddenalpha do you mind updating the description to include the reasoning for this change so that it's more obvious?  Sorry if this is self-explanatory, I just don't completely understand when writing less information than what was instructed is a non-exceptional workflow.
